### PR TITLE
check if a goal is still valid before checking if the goal has changed

### PR DESCRIPTION
**Pull request by u9g**

### CODE REVIEW

**Comment by IceTank** [comment]
I don't think every goal has to implement the function `isValid()` it should work if only the methods that override the default method should implement a `isValid()` method.

**Comment by rom1504** [comment]
I will close this next week if no feedback given

**Comment by IceTank** [comment]
I think the pr is fine the isValid() method is removed from every goal. It should be fine if isValid just returns true by default in the Parent Goal class. Only dynamic goals where the goal could get invalid should need the isValid method.

@u9g Can you refactor it? It should be ready to merge then.

**Comment by rom1504** [comment]
If you want, you can directly edit this @IceTank , you have rights to do it

**Comment by u9g** [comment]
> I think the pr is fine the isValid() method is removed from every goal. It should be fine if isValid just returns true by default in the Parent Goal class. Only dynamic goals where the goal could get invalid should need the isValid method.
> 
> @u9g Can you refactor it? It should be ready to merge then.

Done.

File: lib/goals.js [hunk 1]
```diff
@@ -233,6 +239,13 @@ class GoalCompositeAny extends Goal {
     }
     return false
   }
+
+  isValid () {
+    for (const i in this.goals) {
+      if (!this.goals[i].isValid()) return false
+    }
+    return false
```
Comment (IceTank) [inline lib/goals.js:247]: This should be `return true`

File: lib/goals.js [hunk 2]
```diff
@@ -267,6 +280,13 @@ class GoalCompositeAll extends Goal {
...
```
Comment (IceTank) [inline lib/goals.js:288]: Same here

**Review by IceTank — CHANGES_REQUESTED** [review]

File: lib/goals.js (hunk 1 above)
Comment (u9g) [reply lib/goals.js]: done!

**Review by IceTank — APPROVED** [review]

**Comment by IceTank** [comment]
@Karang anything to add? Or can we merge this?

**Pull request merged**